statistics/histogram.go

```diff
@@ -857,6 +857,202 @@ func (idx *Index) getRowCount(sc *stmtctx.StatementContext, indexRanges []*range
 	return totalCount, nil
 }
 
+func (c *Column) newNumericColumnBySelectivity(sc *stmtctx.StatementContext, statsNode *StatsNode) (*Column, error) {
+	newColHist := &Column{Info: c.Info}
+	newColHist.Histogram = *NewHistogram(c.ID, int64(float64(c.NDV)*statsNode.Selectivity), 0, 0, c.Tp, chunk.InitialCapacity, 0)
+	ranLowIdx := 0
+	var totCnt int64 = 0
+	// Process each bucket.
+	for i := 0; i < c.Bounds.NumRows() && ranLowIdx < len(statsNode.Ranges); i += 2 {
+		newBkt := Bucket{Repeat: c.Buckets[i/2].Repeat}
+		// Find the bucket which is the first one whose have intersection with the bucket.
+		for ; ranLowIdx < len(statsNode.Ranges); ranLowIdx++ {
+			if chunk.Compare(c.Bounds.GetRow(i), 0, &statsNode.Ranges[ranLowIdx].HighVal[0]) <= 0 {
+				break
+			}
+		}
+		ranHighIdx := ranLowIdx
+		// Find the bucket which is the first one whose lowVal is bigger than bucket's high bound, i.e. out of the bucket.
+		for ; ranHighIdx < len(statsNode.Ranges); ranHighIdx++ {
+			if chunk.Compare(c.Bounds.GetRow(i+1), 0, &statsNode.Ranges[ranHighIdx].LowVal[0]) < 0 {
+				break
+			}
+		}
+		if ranLowIdx == ranHighIdx {
+			continue
+		}
+		overlapped := 0.0
+		// Compute the overlap ratio.
+		for ranIdx := ranLowIdx; ranIdx < ranHighIdx; ranIdx++ {
+			overlapped += c.CalcRangeFraction(i/2, &statsNode.Ranges[ranIdx].LowVal[0], &statsNode.Ranges[ranIdx].HighVal[0])
```

Comment:
Can we only use the ranges, not fractions, to decide if they are overlapped? In this way, we can merge the logic of cutting histogram of three cases into one function, and adjust the count of buckets after the histogram is built. We can spare lots of code in this way.

Reply:
How to decide the overlap ratio to calculate the count of new histogram?
I need some more explanation to get your point 😂